### DISCUSSION TITLE
Solving Inverse Problems for Electromagnetic Wave Propagation using MFEM

### DISCUSSION THREAD

**Discussion by cyasing** [mfem/mfem#4314] [Q&A]
I am trying to program for an inverse solver to reconstruct the region's permittivity using example 25 (electromagnetics) as the forward solver. To do that, I have a few questions:

1. To set up the correct system `Ax=b`, I need to evaluate the term: `x^T M x`, where `M` corresponds to the term `\omega^2 * epsilon` in the equation that the example solves. I also need to assemble matrices based on parts of `M`, like a matrix: `[-Mi -Mr; Mr -Mi]` (Mi and Mr are imaginary and real parts of the matrix M) for each element of the region. What would be a safe way of doing this evaluation? I found (#2238) that addresses this, but I would like to know if the custom matrix building has been improved since then.
2. I don't know how to construct the variable operator term: `\omega^2 * epsilon E` that updates each time with a new forward solve for reconstruction from an initial guess. I have set up a `FunctionCoefficient` for spatial change in permittivity, but I don't know how to update it with each iteration. More importantly, since I would need a spatial profile of the term as unknowns for the inverse problem, how I make sure I have values of all locations being updated at each iteration?
3. I am also wondering if I should just use MFEM to solve the inverse system, or should I use another library like `dlib` for the inverse optimization process. If I can just do it in MFEM, what types of solvers would be recommended for the inversion of the same system as in example 25? Are there suitable Conjugate Gradient and quasi-Newton solvers available in MFEM?
4. I found an earlier issue regarding setting up multiple RHS system (#3318), each corresponding to a singular source. Has that functionality improved or will I still need to solve for all sources in a for loop?

I realise this is quite a lot of points, and I can only say I will be really grateful if anyone can answer these doubts as soon as they can, even one at a time would be really valuable.

**Comment by DylanPrinsloo** [comment]
To my knowledge, there is no Quasi-Newton solvers which are readily available in MFEM. Id use a reference manual or a prescribed text from a possible research institution. 

I'm interest to see the solution, so do feel free to keep me/ us updated!

**Reply by cyasing** [reply]
Alright, I will discuss any solutions that will be successful.
